File: java project/project/services/ComplaintService.java
package project.services;

import java.time.LocalDate;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Scanner;
import project.models.Complaint;
import project.utils.FileManager;
import project.utils.InputValidator;
import project.utils.UniqueIDGenerator;

/**
 * Handles all complaint-related operations.
 * This class uses file handling, searching, sorting, queue, and priority queue logic.
 */
public class ComplaintService {

    private static final String COMPLAINT_FILE = "project/data/complaints.txt";
    private static final List<String> DEPARTMENTS = List.of(
            "Electricity", "Road", "Water", "Garbage", "Police",
            "Transport", "Education", "Health", "Others"
    );

    /**
     * Creates the service and ensures the complaint file exists.
     */
    public ComplaintService() {
        FileManager.ensureFileExists(COMPLAINT_FILE);
    }

    /**
     * Loads all complaints from complaints.txt.
     *
     * @return list of complaints
     */
    public List<Complaint> loadComplaints() {
        List<Complaint> complaints = new ArrayList<>();
        List<String> lines = FileManager.readAllLines(COMPLAINT_FILE);

        for (String line : lines) {
            Complaint complaint = Complaint.fromFileString(line);
            if (complaint != null) {
                complaints.add(complaint);
            }
        }
        return complaints;
    }

    /**
     * Saves all complaints back to the file.
     *
     * @param complaints complaint list to save
     */
    public void saveComplaints(List<Complaint> complaints) {
        List<String> lines = new ArrayList<>();
        for (Complaint complaint : complaints) {
            lines.add(complaint.toFileString());
        }
        FileManager.writeAllLines(COMPLAINT_FILE, lines);
    }

    /**
     * Searches for a complaint using complaint ID.
     * This is a linear search.
     *
     * @param complaintId complaint ID to search
     * @return matching complaint or null
     */
    public Complaint findComplaintById(String complaintId) {
        if (InputValidator.isEmpty(complaintId)) {
            return null;
        }

        for (Complaint complaint : loadComplaints()) {
            if (complaint.getComplaintId().equalsIgnoreCase(complaintId.trim())) {
                return complaint;
            }
        }
        return null;
    }

    /**
     * Gets all complaints submitted by a specific username.
     *
     * @param username username to search
     * @return user complaints
     */
    public List<Complaint> getComplaintsByUsername(String username) {
        List<Complaint> result = new ArrayList<>();
        if (InputValidator.isEmpty(username)) {
            return result;
        }

        for (Complaint complaint : loadComplaints()) {
            if (complaint.getUsername().equalsIgnoreCase(username.trim())) {
                result.add(complaint);
            }
        }
        return result;
    }

    /**
     * Gets complaints filtered by status.
     *
     * @param status complaint status
     * @return filtered list
     */
    public List<Complaint> getComplaintsByStatus(String status) {
        List<Complaint> result = new ArrayList<>();
        if (InputValidator.isEmpty(status)) {
            return result;
        }

        for (Complaint complaint : loadComplaints()) {
            if (complaint.getStatus().equalsIgnoreCase(status.trim())) {
                result.add(complaint);
            }
        }
        return result;
    }

    /**
     * Submits a new complaint for the logged-in user.
     *
     * @param scanner scanner for console input
     * @param username logged-in username
     * @return created complaint, or null if submission fails
     */
    public Complaint submitComplaint(Scanner scanner, String username) {
        if (InputValidator.isEmpty(username)) {
            System.out.println("Invalid user.");
            return null;
        }

        System.out.print("Enter Topic: ");
        String topic = scanner.nextLine().trim();

        System.out.println("Choose Priority:");
        System.out.println("1. Low");
        System.out.println("2. Medium");
        System.out.println("3. High");
        System.out.print("Enter choice: ");
        String priorityChoice = scanner.nextLine().trim();

        System.out.print("Enter Description: ");
        String description = scanner.nextLine().trim();

        System.out.print("Enter Location: ");
        String location = scanner.nextLine().trim();

        System.out.println("Choose Department:");
        for (int i = 0; i < DEPARTMENTS.size(); i++) {
            System.out.println((i + 1) + ". " + DEPARTMENTS.get(i));
        }
        System.out.print("Enter choice: ");
        String departmentChoice = scanner.nextLine().trim();

        if (InputValidator.isEmpty(topic) || InputValidator.isEmpty(description) || InputValidator.isEmpty(location)) {
            System.out.println("All fields are required.");
            return null;
        }

        String priority = getPriorityFromChoice(priorityChoice);
        if (priority == null) {
            System.out.println("Invalid priority choice.");
            return null;
        }

        String department = getDepartmentFromChoice(departmentChoice);
        if (department == null) {
            System.out.println("Invalid department choice.");
            return null;
        }

        String complaintId = UniqueIDGenerator.generateComplaintId();
        String date = LocalDate.now().toString();
        String time = LocalTime.now().format(DateTimeFormatter.ofPattern("HH:mm:ss"));
        String status = "Pending";
        String adminRemark = "Not Reviewed";

        Complaint complaint = new Complaint(complaintId, username.trim(), topic, priority, description,
                location, department, date, time, status, adminRemark);

        FileManager.appendLine(COMPLAINT_FILE, complaint.toFileString());
        System.out.println("Complaint Submitted Successfully");
        System.out.println("Complaint ID : " + complaintId);
        return complaint;
    }

    /**
     * Updates only a pending complaint.
     *
     * @param complaintId complaint ID to edit
     * @param scanner console scanner
     * @return true if updated successfully
     */
    public boolean editPendingComplaint(String complaintId, Scanner scanner) {
        List<Complaint> complaints = loadComplaints();
        boolean updated = false;

        for (Complaint complaint : complaints) {
            if (complaint.getComplaintId().equalsIgnoreCase(complaintId.trim())) {
                if (!"Pending".equalsIgnoreCase(complaint.getStatus())) {
                    System.out.println("Only pending complaints can be edited.");
                    return false;
                }

                System.out.print("Enter New Topic: ");
                String topic = scanner.nextLine().trim();

                System.out.println("Choose New Priority:");
                System.out.println("1. Low");
                System.out.println("2. Medium");
                System.out.println("3. High");
                System.out.print("Enter choice: ");
                String priorityChoice = scanner.nextLine().trim();

                System.out.print("Enter New Description: ");
                String description = scanner.nextLine().trim();

                System.out.print("Enter New Location: ");
                String location = scanner.nextLine().trim();

                System.out.println("Choose New Department:");
                for (int i = 0; i < DEPARTMENTS.size(); i++) {
                    System.out.println((i + 1) + ". " + DEPARTMENTS.get(i));
                }
                System.out.print("Enter choice: ");
                String departmentChoice = scanner.nextLine().trim();

                String priority = getPriorityFromChoice(priorityChoice);
                String department = getDepartmentFromChoice(departmentChoice);

                if (InputValidator.isEmpty(topic) || InputValidator.isEmpty(description) || InputValidator.isEmpty(location)
                        || priority == null || department == null) {
                    System.out.println("Invalid input. Complaint not updated.");
                    return false;
                }

                complaint.setTopic(topic);
                complaint.setPriority(priority);
                complaint.setDescription(description);
                complaint.setLocation(location);
                complaint.setDepartment(department);
                updated = true;
                break;
            }
        }

        if (updated) {
            saveComplaints(complaints);
            System.out.println("Complaint updated successfully.");
        } else {
            System.out.println("Complaint not found.");
        }
        return updated;
    }

    /**
     * Deletes a complaint only if it is still pending.
     *
     * @param complaintId complaint ID to delete
     * @return true if deleted
     */
    public boolean deletePendingComplaint(String complaintId) {
        List<Complaint> complaints = loadComplaints();
        List<Complaint> updatedList = new ArrayList<>();
        boolean deleted = false;

        for (Complaint complaint : complaints) {
            if (complaint.getComplaintId().equalsIgnoreCase(complaintId.trim())) {
                if (!"Pending".equalsIgnoreCase(complaint.getStatus())) {
                    System.out.println("Only pending complaints can be deleted.");
                    return false;
                }
                deleted = true;
                continue;
            }
            updatedList.add(complaint);
        }

        if (deleted) {
            saveComplaints(updatedList);
            System.out.println("Complaint deleted successfully.");
        } else {
            System.out.println("Complaint not found.");
        }
        return deleted;
    }

    /**
     * Updates status for a complaint.
     *
     * @param complaintId complaint ID
     * @param newStatus new status value
     * @return true if updated
     */
    public boolean updateComplaintStatus(String complaintId, String newStatus) {
        List<Complaint> complaints = loadComplaints();
        boolean updated = false;

        for (Complaint complaint : complaints) {
            if (complaint.getComplaintId().equalsIgnoreCase(complaintId.trim())) {
                complaint.setStatus(newStatus);
                updated = true;
                break;
            }
        }

        if (updated) {
            saveComplaints(complaints);
        }
        return updated;
    }

    /**
     * Updates the admin remark for a complaint.
     *
     * @param complaintId complaint ID
     * @param remark admin remark text
     * @return true if updated
     */
    public boolean updateAdminRemark(String complaintId, String remark) {
        List<Complaint> complaints = loadComplaints();
        boolean updated = false;

        for (Complaint complaint : complaints) {
            if (complaint.getComplaintId().equalsIgnoreCase(complaintId.trim())) {
                complaint.setAdminRemark(remark);
                updated = true;
                break;
            }
        }

        if (updated) {
            saveComplaints(complaints);
        }
        return updated;
    }

    /**
     * Returns all complaints sorted by a selected field.
     *
     * @param sortBy field name
     * @return sorted complaints
     */
    public List<Complaint> sortComplaints(String sortBy) {
        List<Complaint> complaints = loadComplaints();

        Comparator<Complaint> comparator;
        if ("priority".equalsIgnoreCase(sortBy)) {
            comparator = Comparator.comparingInt(c -> getPriorityRank(c.getPriority()));
        } else if ("date".equalsIgnoreCase(sortBy)) {
            comparator = Comparator.comparing(Complaint::getDate).thenComparing(Complaint::getTime);
        } else if ("department".equalsIgnoreCase(sortBy)) {
            comparator = Comparator.comparing(Complaint::getDepartment, String.CASE_INSENSITIVE_ORDER);
        } else if ("status".equalsIgnoreCase(sortBy)) {
            comparator = Comparator.comparing(Complaint::getStatus, String.CASE_INSENSITIVE_ORDER);
        } else {
            comparator = Comparator.comparing(Complaint::getComplaintId, String.CASE_INSENSITIVE_ORDER);
        }

        Collections.sort(complaints, comparator);
        return complaints;
    }

    /**
     * Returns pending complaints in a queue.
     *
     * @return queue of pending complaints
     */
    public Queue<Complaint> getPendingQueue() {
        Queue<Complaint> queue = new LinkedList<>();
        for (Complaint complaint : loadComplaints()) {
            if ("Pending".equalsIgnoreCase(complaint.getStatus())) {
                queue.offer(complaint);
            }
        }
        return queue;
    }

    /**
     * Returns complaints in priority queue order.
     * Highest priority is processed first.
     *
     * @return priority queue of complaints
     */
    public PriorityQueue<Complaint> getPriorityQueue() {
        PriorityQueue<Complaint> queue = new PriorityQueue<>(Comparator
                .comparingInt((Complaint c) -> getPriorityRank(c.getPriority())));

        queue.addAll(loadComplaints());
        return queue;
    }

    /**
     * Returns dashboard statistics.
     *
     * @return map of summary counts
     */
    public Map<String, Integer> getDashboardStatistics() {
        List<Complaint> complaints = loadComplaints();
        Map<String, Integer> stats = new HashMap<>();

        stats.put("Total", complaints.size());
        stats.put("Pending", 0);
        stats.put("Accepted", 0);
        stats.put("Resolved", 0);
        stats.put("Rejected", 0);
        stats.put("High", 0);
        stats.put("Medium", 0);
        stats.put("Low", 0);

        for (Complaint complaint : complaints) {
            incrementCount(stats, complaint.getStatus());
            incrementCount(stats, complaint.getPriority());
        }

        return stats;
    }

    /**
     * Prints a full complaint detail on the console.
     *
     * @param complaint complaint to display
     */
    public void printComplaintDetails(Complaint complaint) {
        if (complaint == null) {
            System.out.println("Complaint Not Found");
            return;
        }

        System.out.println("--------------------------------------------------");
        System.out.println("Complaint ID   : " + complaint.getComplaintId());
        System.out.println("Username       : " + complaint.getUsername());
        System.out.println("Topic          : " + complaint.getTopic());
        System.out.println("Priority       : " + complaint.getPriority());
        System.out.println("Description    : " + complaint.getDescription());
        System.out.println("Location       : " + complaint.getLocation());
        System.out.println("Department     : " + complaint.getDepartment());
        System.out.println("Date           : " + complaint.getDate());
        System.out.println("Time           : " + complaint.getTime());
        System.out.println("Status         : " + complaint.getStatus());
        System.out.println("Admin Remark   : " + complaint.getAdminRemark());
        System.out.println("--------------------------------------------------");
    }

    /**
     * Converts a priority selection into text.
     *
     * @param choice menu choice
     * @return Low, Medium, High, or null
     */
    private String getPriorityFromChoice(String choice) {
        if ("1".equals(choice)) {
            return "Low";
        }
        if ("2".equals(choice)) {
            return "Medium";
        }
        if ("3".equals(choice)) {
            return "High";
        }
        return null;
    }

    /**
     * Converts a department selection into text.
     *
     * @param choice menu choice
     * @return department name or null
     */
    private String getDepartmentFromChoice(String choice) {
        try {
            int index = Integer.parseInt(choice);
            if (index >= 1 && index <= DEPARTMENTS.size()) {
                return DEPARTMENTS.get(index - 1);
            }
        } catch (NumberFormatException ignored) {
            // invalid choice, handled by returning null
        }
        return null;
    }

    /**
     * Returns a numeric rank for priority sorting.
     * High comes first, then Medium, then Low.
     *
     * @param priority priority text
     * @return rank number
     */
    private int getPriorityRank(String priority) {
        if ("High".equalsIgnoreCase(priority)) {
            return 1;
        }
        if ("Medium".equalsIgnoreCase(priority)) {
            return 2;
        }
        return 3;
    }

    /**
     * Increments a count in the statistics map.
     *
     * @param stats stats map
     * @param key key to increment
     */
    private void incrementCount(Map<String, Integer> stats, String key) {
        if (key == null) {
            return;
        }

        if (stats.containsKey(key)) {
            stats.put(key, stats.get(key) + 1);
        }
    }
}

File: cpp_project/services/ComplaintService.cpp
#include "ComplaintService.h"
#include "../utils/FileManager.h"
#include "../utils/InputValidator.h"
#include "../utils/UniqueIDGenerator.h"
#include <iostream>
#include <algorithm>
#include <chrono>
#include <ctime>
#include <iomanip>
#include <sstream>

namespace project {
namespace services {

ComplaintService::ComplaintService() {
    utils::FileManager::ensureFileExists(COMPLAINT_FILE);
}

std::vector<models::Complaint> ComplaintService::loadComplaints() {
    std::vector<models::Complaint> complaints;
    std::vector<std::string> lines = utils::FileManager::readAllLines(COMPLAINT_FILE);

    for (const std::string& line : lines) {
        models::Complaint complaint = models::Complaint::fromFileString(line);
        if (complaint.isValid()) {
            complaints.push_back(complaint);
        }
    }
    return complaints;
}

void ComplaintService::saveComplaints(const std::vector<models::Complaint>& complaints) {
    std::vector<std::string> lines;
    for (const models::Complaint& complaint : complaints) {
        lines.push_back(complaint.toFileString());
    }
    utils::FileManager::writeAllLines(COMPLAINT_FILE, lines);
}

models::Complaint ComplaintService::findComplaintById(const std::string& complaintId) {
    if (utils::InputValidator::isEmpty(complaintId)) return models::Complaint();

    std::vector<models::Complaint> complaints = loadComplaints();
    for (const models::Complaint& complaint : complaints) {
        if (complaint.getComplaintId() == complaintId) {
            return complaint;
        }
    }
    return models::Complaint();
}

std::vector<models::Complaint> ComplaintService::getComplaintsByUsername(const std::string& username) {
    std::vector<models::Complaint> result;
    if (utils::InputValidator::isEmpty(username)) return result;

    std::vector<models::Complaint> complaints = loadComplaints();
    for (const models::Complaint& complaint : complaints) {
        if (complaint.getUsername() == username) {
            result.push_back(complaint);
        }
    }
    return result;
}

std::vector<models::Complaint> ComplaintService::getComplaintsByStatus(const std::string& status) {
    std::vector<models::Complaint> result;
    if (utils::InputValidator::isEmpty(status)) return result;

    std::vector<models::Complaint> complaints = loadComplaints();
    for (const models::Complaint& complaint : complaints) {
        if (complaint.getStatus() == status) {
            result.push_back(complaint);
        }
    }
    return result;
}

models::Complaint ComplaintService::submitComplaint(const std::string& username) {
    if (utils::InputValidator::isEmpty(username)) {
        std::cout << "Invalid user.\n";
        return models::Complaint();
    }

    std::string topic, priorityChoice, description, location, departmentChoice;

    std::cout << "Enter Topic: ";
    std::getline(std::cin, topic);

    std::cout << "Choose Priority:\n1. Low\n2. Medium\n3. High\nEnter choice: ";
    std::getline(std::cin, priorityChoice);

    std::cout << "Enter Description: ";
    std::getline(std::cin, description);

    std::cout << "Enter Location: ";
    std::getline(std::cin, location);

    std::cout << "Choose Department:\n";
    for (size_t i = 0; i < DEPARTMENTS.size(); i++) {
        std::cout << (i + 1) << ". " << DEPARTMENTS[i] << "\n";
    }
    std::cout << "Enter choice: ";
    std::getline(std::cin, departmentChoice);

    if (utils::InputValidator::isEmpty(topic) || utils::InputValidator::isEmpty(description) || utils::InputValidator::isEmpty(location)) {
        std::cout << "All fields are required.\n";
        return models::Complaint();
    }

    std::string priority = getPriorityFromChoice(priorityChoice);
    if (priority.empty()) {
        std::cout << "Invalid priority choice.\n";
        return models::Complaint();
    }

    std::string department = getDepartmentFromChoice(departmentChoice);
    if (department.empty()) {
        std::cout << "Invalid department choice.\n";
        return models::Complaint();
    }

    std::string complaintId = utils::UniqueIDGenerator::generateComplaintId();
    
    auto now = std::chrono::system_clock::now();
    std::time_t now_time = std::chrono::system_clock::to_time_t(now);
    std::tm* local_time = std::localtime(&now_time);
    
    std::stringstream date_ss, time_ss;
    date_ss << std::put_time(local_time, "%Y-%m-%d");
    time_ss << std::put_time(local_time, "%H:%M:%S");

    models::Complaint complaint(complaintId, username, topic, priority, description, location,
                                department, date_ss.str(), time_ss.str(), "Pending", "Not Reviewed");

    utils::FileManager::appendLine(COMPLAINT_FILE, complaint.toFileString());
    std::cout << "Complaint Submitted Successfully\n";
    std::cout << "Complaint ID : " << complaintId << "\n";
    return complaint;
}

bool ComplaintService::editPendingComplaint(const std::string& complaintId) {
    std::vector<models::Complaint> complaints = loadComplaints();
    bool updated = false;

    for (models::Complaint& complaint : complaints) {
        if (complaint.getComplaintId() == complaintId) {
            if (complaint.getStatus() != "Pending") {
                std::cout << "Only pending complaints can be edited.\n";
                return false;
            }

            std::string topic, priorityChoice, description, location, departmentChoice;

            std::cout << "Enter New Topic: ";
            std::getline(std::cin, topic);

            std::cout << "Choose New Priority:\n1. Low\n2. Medium\n3. High\nEnter choice: ";
            std::getline(std::cin, priorityChoice);

            std::cout << "Enter New Description: ";
            std::getline(std::cin, description);

            std::cout << "Enter New Location: ";
            std::getline(std::cin, location);

            std::cout << "Choose New Department:\n";
            for (size_t i = 0; i < DEPARTMENTS.size(); i++) {
                std::cout << (i + 1) << ". " << DEPARTMENTS[i] << "\n";
            }
            std::cout << "Enter choice: ";
            std::getline(std::cin, departmentChoice);

            std::string priority = getPriorityFromChoice(priorityChoice);
            std::string department = getDepartmentFromChoice(departmentChoice);

            if (utils::InputValidator::isEmpty(topic) || utils::InputValidator::isEmpty(description) || 
                utils::InputValidator::isEmpty(location) || priority.empty() || department.empty()) {
                std::cout << "Invalid input. Complaint not updated.\n";
                return false;
            }

            complaint.setTopic(topic);
            complaint.setPriority(priority);
            complaint.setDescription(description);
            complaint.setLocation(location);
            complaint.setDepartment(department);
            updated = true;
            break;
        }
    }

    if (updated) {
        saveComplaints(complaints);
        std::cout << "Complaint updated successfully.\n";
    } else {
        std::cout << "Complaint not found.\n";
    }
    return updated;
}

bool ComplaintService::deletePendingComplaint(const std::string& complaintId) {
    std::vector<models::Complaint> complaints = loadComplaints();
    std::vector<models::Complaint> updatedList;
    bool deleted = false;

    for (const models::Complaint& complaint : complaints) {
        if (complaint.getComplaintId() == complaintId) {
            if (complaint.getStatus() != "Pending") {
                std::cout << "Only pending complaints can be deleted.\n";
                return false;
            }
            deleted = true;
            continue;
        }
        updatedList.push_back(complaint);
    }

    if (deleted) {
        saveComplaints(updatedList);
        std::cout << "Complaint deleted successfully.\n";
    } else {
        std::cout << "Complaint not found.\n";
    }
    return deleted;
}

bool ComplaintService::updateComplaintStatus(const std::string& complaintId, const std::string& newStatus) {
    std::vector<models::Complaint> complaints = loadComplaints();
    bool updated = false;

    for (models::Complaint& complaint : complaints) {
        if (complaint.getComplaintId() == complaintId) {
            complaint.setStatus(newStatus);
            updated = true;
            break;
        }
    }

    if (updated) saveComplaints(complaints);
    return updated;
}

bool ComplaintService::updateAdminRemark(const std::string& complaintId, const std::string& remark) {
    std::vector<models::Complaint> complaints = loadComplaints();
    bool updated = false;

    for (models::Complaint& complaint : complaints) {
        if (complaint.getComplaintId() == complaintId) {
            complaint.setAdminRemark(remark);
            updated = true;
            break;
        }
    }

    if (updated) saveComplaints(complaints);
    return updated;
}

std::vector<models::Complaint> ComplaintService::sortComplaints(const std::string& sortBy) {
    std::vector<models::Complaint> complaints = loadComplaints();

    if (sortBy == "priority") {
        std::sort(complaints.begin(), complaints.end(), [this](const models::Complaint& a, const models::Complaint& b) {
            return getPriorityRank(a.getPriority()) < getPriorityRank(b.getPriority());
        });
    } else if (sortBy == "date") {
        std::sort(complaints.begin(), complaints.end(), [](const models::Complaint& a, const models::Complaint& b) {
            if (a.getDate() == b.getDate()) return a.getTime() < b.getTime();
            return a.getDate() < b.getDate();
        });
    } else if (sortBy == "department") {
        std::sort(complaints.begin(), complaints.end(), [](const models::Complaint& a, const models::Complaint& b) {
            return a.getDepartment() < b.getDepartment();
        });
    } else if (sortBy == "status") {
        std::sort(complaints.begin(), complaints.end(), [](const models::Complaint& a, const models::Complaint& b) {
            return a.getStatus() < b.getStatus();
        });
    } else {
        std::sort(complaints.begin(), complaints.end(), [](const models::Complaint& a, const models::Complaint& b) {
            return a.getComplaintId() < b.getComplaintId();
        });
    }
    
    return complaints;
}

std::queue<models::Complaint> ComplaintService::getPendingQueue() {
    std::queue<models::Complaint> q;
    std::vector<models::Complaint> complaints = loadComplaints();
    for (const models::Complaint& c : complaints) {
        if (c.getStatus() == "Pending") {
            q.push(c);
        }
    }
    return q;
}

std::map<std::string, int> ComplaintService::getDashboardStatistics() {
    std::vector<models::Complaint> complaints = loadComplaints();
    std::map<std::string, int> stats = {
        {"Total", (int)complaints.size()},
        {"Pending", 0}, {"Accepted", 0}, {"Resolved", 0}, {"Rejected", 0},
        {"High", 0}, {"Medium", 0}, {"Low", 0}
    };

    for (const models::Complaint& c : complaints) {
        incrementCount(stats, c.getStatus());
        incrementCount(stats, c.getPriority());
    }

    return stats;
}

void ComplaintService::printComplaintDetails(const models::Complaint& complaint) {
    if (!complaint.isValid()) {
        std::cout << "Complaint Not Found\n";
        return;
    }

    std::cout << "--------------------------------------------------\n";
    std::cout << "Complaint ID   : " << complaint.getComplaintId() << "\n";
    std::cout << "Username       : " << complaint.getUsername() << "\n";
    std::cout << "Topic          : " << complaint.getTopic() << "\n";
    std::cout << "Priority       : " << complaint.getPriority() << "\n";
    std::cout << "Description    : " << complaint.getDescription() << "\n";
    std::cout << "Location       : " << complaint.getLocation() << "\n";
    std::cout << "Department     : " << complaint.getDepartment() << "\n";
    std::cout << "Date           : " << complaint.getDate() << "\n";
    std::cout << "Time           : " << complaint.getTime() << "\n";
    std::cout << "Status         : " << complaint.getStatus() << "\n";
    std::cout << "Admin Remark   : " << complaint.getAdminRemark() << "\n";
    std::cout << "--------------------------------------------------\n";
}

std::string ComplaintService::getPriorityFromChoice(const std::string& choice) {
    if (choice == "1") return "Low";
    if (choice == "2") return "Medium";
    if (choice == "3") return "High";
    return "";
}

std::string ComplaintService::getDepartmentFromChoice(const std::string& choice) {
    try {
        int idx = std::stoi(choice);
        if (idx >= 1 && idx <= (int)DEPARTMENTS.size()) {
            return DEPARTMENTS[idx - 1];
        }
    } catch (...) {}
    return "";
}

int ComplaintService::getPriorityRank(const std::string& priority) const {
    if (priority == "High") return 1;
    if (priority == "Medium") return 2;
    return 3;
}

void ComplaintService::incrementCount(std::map<std::string, int>& stats, const std::string& key) {
    if (stats.find(key) != stats.end()) {
        stats[key]++;
    }
}

}
}
